File: server/services/tag_alias_service.py
from __future__ import annotations

import json
import logging
from dataclasses import dataclass
from pathlib import Path


logger = logging.getLogger(__name__)


def _normalize_name(name: str) -> str:
    return name.strip().casefold()


def _lookup_key(category: str, name: str) -> str:
    return f"{category}\0{_normalize_name(name)}"


@dataclass(frozen=True)
class _TagAliasGroup:
    category: str
    canonical: str
    aliases: list[str]
    all_names: list[str]

    @classmethod
    def create(
        cls,
        *,
        category: str,
        canonical: str,
        aliases: list[str],
    ) -> "_TagAliasGroup | None":
        normalized_canonical = canonical.strip()
        if not normalized_canonical:
            return None

        out: list[str] = []
        seen: set[str] = set()

        def add_value(value: str) -> None:
            trimmed = value.strip()
            if not trimmed:
                return
            normalized = _normalize_name(trimmed)
            if normalized in seen:
                return
            seen.add(normalized)
            out.append(trimmed)

        add_value(normalized_canonical)
        for alias in aliases:
            add_value(alias)

        if not out:
            return None

        return cls(
            category=category,
            canonical=normalized_canonical,
            aliases=[value for value in out if value != normalized_canonical],
            all_names=out,
        )

    def matches_partial(self, raw_query: str) -> bool:
        normalized_query = _normalize_name(raw_query)
        if not normalized_query:
            return True
        return any(normalized_query in _normalize_name(value) for value in self.all_names)


class TagAliasService:
    _supported_categories = ("series", "character")

    def __init__(
        self,
        *,
        groups_by_category: dict[str, list[_TagAliasGroup]] | None = None,
        groups_by_lookup_key: dict[str, _TagAliasGroup] | None = None,
    ) -> None:
        self._groups_by_category = groups_by_category or {}
        self._groups_by_lookup_key = groups_by_lookup_key or {}

    @classmethod
    def load_default(cls, config_path: Path) -> "TagAliasService":
        try:
            raw_json = config_path.read_text(encoding="utf-8")
        except FileNotFoundError:
            logger.warning("[tag-alias] config missing: %s", config_path)
            return cls()
        except Exception:
            logger.exception("[tag-alias] failed to read config: %s", config_path)
            return cls()

        try:
            return cls.from_json_text(raw_json)
        except Exception:
            logger.exception("[tag-alias] failed to parse config: %s", config_path)
            return cls()

    @classmethod
    def from_json_text(cls, raw_json: str) -> "TagAliasService":
        decoded = json.loads(raw_json)
        if not isinstance(decoded, dict):
            return cls()

        groups_by_category: dict[str, list[_TagAliasGroup]] = {}
        groups_by_lookup_key: dict[str, _TagAliasGroup] = {}

        for category in cls._supported_categories:
            category_map = decoded.get(category)
            if not isinstance(category_map, dict):
                continue

            groups: list[_TagAliasGroup] = []
            for canonical, raw_aliases in category_map.items():
                aliases: list[str]
                if isinstance(raw_aliases, list):
                    aliases = [str(value) for value in raw_aliases]
                elif isinstance(raw_aliases, str):
                    aliases = [raw_aliases]
                else:
                    aliases = []

                group = _TagAliasGroup.create(
                    category=category,
                    canonical=str(canonical),
                    aliases=aliases,
                )
                if group is None:
                    continue

                groups.append(group)
                for candidate in group.all_names:
                    groups_by_lookup_key[_lookup_key(category, candidate)] = group

            if groups:
                groups_by_category[category] = groups

        return cls(
            groups_by_category=groups_by_category,
            groups_by_lookup_key=groups_by_lookup_key,
        )

    @property
    def is_configured(self) -> bool:
        return bool(self._groups_by_lookup_key)

    def supports_category(self, category: str) -> bool:
        normalized_category = str(category or "").strip()
        return bool(self._groups_by_category.get(normalized_category))

    def canonicalize_name(self, category: str, raw_name: str) -> str:
        trimmed = str(raw_name or "").strip()
        if not trimmed:
            return ""
        group = self._groups_by_lookup_key.get(_lookup_key(category, trimmed))
        return group.canonical if group is not None else trimmed

    def equivalent_names(
        self,
        category: str,
        raw_name: str,
        *,
        partial: bool = False,
    ) -> list[str]:
        trimmed = str(raw_name or "").strip()
        if not trimmed:
            return []

        normalized_category = str(category or "").strip()
        groups = self._groups_by_category.get(normalized_category) or []
        out: list[str] = []
        seen: set[str] = set()

        def add_candidate(value: str) -> None:
            normalized = value.strip()
            if not normalized:
                return
            lookup = _normalize_name(normalized)
            if lookup in seen:
                return
            seen.add(lookup)
            out.append(normalized)

        if not groups:
            add_candidate(trimmed)
            return out

        if partial:
            for group in groups:
                if not group.matches_partial(trimmed):
                    continue
                for candidate in group.all_names:
                    add_candidate(candidate)
            add_candidate(trimmed)
            return out

        group = self._groups_by_lookup_key.get(_lookup_key(normalized_category, trimmed))
        if group is not None:
            for candidate in group.all_names:
                add_candidate(candidate)
            return out

        add_candidate(trimmed)
        return out

    def equivalent_names_across_categories(
        self,
        raw_name: str,
        *,
        partial: bool = False,
    ) -> list[str]:
        trimmed = str(raw_name or "").strip()
        if not trimmed:
            return []

        out: list[str] = []
        seen: set[str] = set()

        def add_candidate(value: str) -> None:
            normalized = value.strip()
            if not normalized:
                return
            lookup = _normalize_name(normalized)
            if lookup in seen:
                return
            seen.add(lookup)
            out.append(normalized)

        for category in self._supported_categories:
            for candidate in self.equivalent_names(category, trimmed, partial=partial):
                add_candidate(candidate)

        if not out:
            add_candidate(trimmed)
        return out

    def matches_contains(self, category: str, candidate_name: str, contains: str) -> bool:
        normalized_contains = _normalize_name(contains)
        if not normalized_contains:
            return True

        group = self._groups_by_lookup_key.get(_lookup_key(category, candidate_name))
        if group is None:
            return normalized_contains in _normalize_name(candidate_name)

        return any(normalized_contains in _normalize_name(value) for value in group.all_names)

File: lib/services/tag_alias_service.dart
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart';

import '../models/tag.dart';

class TagAliasService {
  static const String _assetPath = 'assets/config/tag_aliases.json';
  static Future<TagAliasService>? _defaultFuture;

  final Map<TagCategory, List<_TagAliasGroup>> _groupsByCategory;
  final Map<String, _TagAliasGroup> _groupsByLookupKey;

  TagAliasService._(
    this._groupsByCategory,
    this._groupsByLookupKey,
  );

  factory TagAliasService.empty() {
    return TagAliasService._(
      <TagCategory, List<_TagAliasGroup>>{},
      <String, _TagAliasGroup>{},
    );
  }

  static Future<TagAliasService> loadDefault() {
    return _defaultFuture ??= _loadDefaultInternal();
  }

  static Future<TagAliasService> _loadDefaultInternal() async {
    try {
      final rawJson = await rootBundle.loadString(_assetPath);
      return TagAliasService.fromJsonString(rawJson);
    } catch (error, stackTrace) {
      debugPrint('[TagAliasService] Failed to load $_assetPath: $error');
      debugPrintStack(
        label: '[TagAliasService] loadDefault',
        stackTrace: stackTrace,
      );
      return TagAliasService.empty();
    }
  }

  factory TagAliasService.fromJsonString(String rawJson) {
    final decoded = jsonDecode(rawJson);
    if (decoded is! Map) {
      return TagAliasService.empty();
    }

    final groupsByCategory = <TagCategory, List<_TagAliasGroup>>{};
    final groupsByLookupKey = <String, _TagAliasGroup>{};

    for (final category in _supportedCategories) {
      final categoryMap = decoded[_categoryKey(category)];
      if (categoryMap is! Map) {
        continue;
      }

      final groups = <_TagAliasGroup>[];
      for (final entry in categoryMap.entries) {
        final rawCanonical = entry.key.toString();
        final aliases = switch (entry.value) {
          List<dynamic> values => values.map((value) => value.toString()).toList(),
          String value => <String>[value],
          _ => const <String>[],
        };
        final group = _TagAliasGroup.create(
          category: category,
          canonical: rawCanonical,
          aliases: aliases,
        );
        if (group == null) {
          continue;
        }
        groups.add(group);
        for (final candidate in group.allNames) {
          groupsByLookupKey[_lookupKey(category, candidate)] = group;
        }
      }

      if (groups.isNotEmpty) {
        groupsByCategory[category] = groups;
      }
    }

    return TagAliasService._(groupsByCategory, groupsByLookupKey);
  }

  bool get isConfigured => _groupsByLookupKey.isNotEmpty;

  bool supportsCategory(TagCategory category) {
    return (_groupsByCategory[category] ?? const <_TagAliasGroup>[]).isNotEmpty;
  }

  String canonicalizeName(TagCategory category, String rawName) {
    final trimmed = rawName.trim();
    if (trimmed.isEmpty) {
      return '';
    }
    final group = _groupsByLookupKey[_lookupKey(category, trimmed)];
    return group?.canonical ?? trimmed;
  }

  Tag canonicalizeTag(Tag tag) {
    final canonicalName = canonicalizeName(tag.category, tag.name);
    return Tag(
      name: canonicalName.isEmpty ? tag.name.trim() : canonicalName,
      category: tag.category,
    );
  }

  List<Tag> canonicalizeTags(Iterable<Tag> tags) {
    final out = <Tag>[];
    final seen = <String>{};
    for (final tag in tags) {
      final canonical = canonicalizeTag(tag);
      final trimmed = canonical.name.trim();
      if (trimmed.isEmpty) {
        continue;
      }
      final key = '${canonical.category.name}\u0000${trimmed.toLowerCase()}';
      if (!seen.add(key)) {
        continue;
      }
      out.add(Tag(name: trimmed, category: canonical.category));
    }
    return out;
  }

  List<String> equivalentNames(
    TagCategory category,
    String rawName, {
    bool partial = false,
  }) {
    final trimmed = rawName.trim();
    if (trimmed.isEmpty) {
      return const <String>[];
    }

    final out = <String>[];
    final seen = <String>{};

    void addCandidate(String value) {
      final normalized = value.trim();
      if (normalized.isEmpty) {
        return;
      }
      final key = normalized.toLowerCase();
      if (seen.add(key)) {
        out.add(normalized);
      }
    }

    if (!supportsCategory(category)) {
      addCandidate(trimmed);
      return out;
    }

    final groups = _groupsByCategory[category] ?? const <_TagAliasGroup>[];
    if (partial) {
      final query = trimmed.toLowerCase();
      for (final group in groups) {
        if (!group.matchesPartial(query)) {
          continue;
        }
        for (final candidate in group.allNames) {
          addCandidate(candidate);
        }
      }
      addCandidate(trimmed);
      return out;
    }

    final group = _groupsByLookupKey[_lookupKey(category, trimmed)];
    if (group != null) {
      for (final candidate in group.allNames) {
        addCandidate(candidate);
      }
      return out;
    }

    addCandidate(trimmed);
    return out;
  }

  List<String> equivalentNamesAcrossAliasCategories(
    String rawName, {
    bool partial = false,
  }) {
    final trimmed = rawName.trim();
    if (trimmed.isEmpty) {
      return const <String>[];
    }

    final out = <String>[];
    final seen = <String>{};

    void addCandidate(String value) {
      final normalized = value.trim();
      if (normalized.isEmpty) {
        return;
      }
      final key = normalized.toLowerCase();
      if (seen.add(key)) {
        out.add(normalized);
      }
    }

    for (final category in _supportedCategories) {
      for (final candidate
          in equivalentNames(category, trimmed, partial: partial)) {
        addCandidate(candidate);
      }
    }

    if (out.isEmpty) {
      addCandidate(trimmed);
    }
    return out;
  }

  bool matchesTagName(
    TagCategory category,
    String candidateName,
    String query, {
    bool partial = false,
  }) {
    final normalizedCandidate = candidateName.trim().toLowerCase();
    if (normalizedCandidate.isEmpty) {
      return false;
    }

    final searchTerms = equivalentNames(category, query, partial: partial);
    if (searchTerms.isEmpty) {
      return false;
    }

    if (partial) {
      return searchTerms.any(
        (term) => normalizedCandidate.contains(term.trim().toLowerCase()),
      );
    }

    return searchTerms.any(
      (term) => normalizedCandidate == term.trim().toLowerCase(),
    );
  }

  bool matchesContains(
    TagCategory category,
    String candidateName,
    String contains,
  ) {
    final normalizedContains = contains.trim().toLowerCase();
    if (normalizedContains.isEmpty) {
      return true;
    }

    final group = _groupsByLookupKey[_lookupKey(category, candidateName)];
    if (group == null) {
      return candidateName.trim().toLowerCase().contains(normalizedContains);
    }

    return group.allNames.any(
      (candidate) => candidate.trim().toLowerCase().contains(normalizedContains),
    );
  }

  static const List<TagCategory> _supportedCategories = <TagCategory>[
    TagCategory.series,
    TagCategory.character,
  ];

  static String _categoryKey(TagCategory category) {
    switch (category) {
      case TagCategory.artist:
        return 'artist';
      case TagCategory.series:
        return 'series';
      case TagCategory.mediaType:
        return 'mediaType';
      case TagCategory.character:
        return 'character';
      case TagCategory.free:
        return 'free';
    }
  }

  static String _lookupKey(TagCategory category, String rawName) {
    return '${category.name}\u0000${rawName.trim().toLowerCase()}';
  }
}

class _TagAliasGroup {
  final TagCategory category;
  final String canonical;
  final List<String> aliases;
  final List<String> allNames;

  const _TagAliasGroup({
    required this.category,
    required this.canonical,
    required this.aliases,
    required this.allNames,
  });

  static _TagAliasGroup? create({
    required TagCategory category,
    required String canonical,
    required List<String> aliases,
  }) {
    final normalizedCanonical = canonical.trim();
    if (normalizedCanonical.isEmpty) {
      return null;
    }

    final out = <String>[];
    final seen = <String>{};

    void addValue(String value) {
      final trimmed = value.trim();
      if (trimmed.isEmpty) {
        return;
      }
      final key = trimmed.toLowerCase();
      if (seen.add(key)) {
        out.add(trimmed);
      }
    }

    addValue(normalizedCanonical);
    for (final alias in aliases) {
      addValue(alias);
    }

    if (out.isEmpty) {
      return null;
    }

    return _TagAliasGroup(
      category: category,
      canonical: normalizedCanonical,
      aliases: out.where((value) => value != normalizedCanonical).toList(),
      allNames: out,
    );
  }

  bool matchesPartial(String normalizedQuery) {
    if (normalizedQuery.isEmpty) {
      return true;
    }
    return allNames.any(
      (candidate) => candidate.trim().toLowerCase().contains(normalizedQuery),
    );
  }
}
